c/tskit/tables.c

```diff
@@ -5341,6 +5334,547 @@ ancestor_mapper_run(ancestor_mapper_t *self)
+            }
+            if (ret != 0) {
+                goto out;
+            }
+
+            // For samples that appear in the parent column of the edge table
+            if (self->is_sample[current_parent] && parent_should_be_added) {
+                ret = ibd_finder_enqueue_segment(
+                    self, 0, self->sequence_length, current_parent);
+                if (ret != 0) {
+                    goto out;
+                }
+                ret = tsk_ibd_finder_calculate_ibd(self, current_parent);
+                if (ret != 0) {
+                    goto out;
+                }
+                parent_should_be_added = false;
+            }
+
+            // if (j != num_edges - 1 && input_edges->parent[j + 1] != current_parent) {
+            //     parent_should_be_added = true;
+            //     for (k = 0; k < self->num_nodes; k++) {
+            //         // This feels rly inefficient now... a better way?
+            //         if (current_parent == self->oldest_parent[k]) {
+            //             // The above causes a segfault at present.
+            //             printf("TODO: Memory can be freed here, think about
+            //             later...\n");
+            //         }
+            //     }
+            // }
+        }
+    }
+
+out:
+    return ret;
+}
+
+int TSK_WARN_UNUSED
+tsk_ibd_finder_calculate_ibd(tsk_ibd_finder_t *self, tsk_id_t current_parent)
+{
+    int ret = 0;
+    int j;
+    tsk_segment_t *seg, *seg0, *seg1;
+    double l, r;
+
+    if (self->ancestor_map_head[current_parent] == NULL) {
+        for (j = 0; j != (int) self->segment_queue_size; j++) {
+            seg = &self->segment_queue[j];
+            ret = ibd_finder_add_ancestry(
+                self, current_parent, seg->left, seg->right, seg->node);
+            if (ret != 0) {
+                goto out;
+            }
+        }
+    } else {
+        for (seg0 = self->ancestor_map_head[current_parent]; seg0 != NULL;
+             seg0 = seg0->next) {
+            for (j = 0; j != (int) self->segment_queue_size; j++) {
+                seg1 = &self->segment_queue[j];
+                if (seg0->left > seg1->left) {
+                    l = seg0->left;
+                } else {
+                    l = seg1->left;
+                }
+                if (seg0->right < seg1->right) {
+                    r = seg0->right;
+                } else {
+                    r = seg0->right;
+                }
+
+                if (l < r) {
+                    if (r - l > self->min_length) {
+                        if (seg0->node < seg1->node) {
+                            ret = ibd_finder_find_sample_pair_index(
+                                self, seg0->node, seg1->node);
+                        } else {
+                            ret = ibd_finder_find_sample_pair_index(
+                                self, seg1->node, seg0->node);
+                        }
+                        if (ret != 0) {
```

Comment:
So, here we'd do
```c
if (ret < 0) {
    goto out;
}
pair_index = ret;
 ret = ibd_finder_add_output(elf, l, r, current_parent, pair_index);

Reply:
Hmm, although it's not clear to me how we'll map back from arbitrary sample pairs to their indexes in the input list here. It's probably simplest to do a linear scan of the list of pairs for now, and come with something better later.

Reply:
Yeah, this is something I didn't think of in the call -- the function that defines the mapping between the 2D sample pairs and the ID index is going to be a lot more complicated in this new setting of arbitrary sample pairs. You could still come up with a canonical way of ordering them, but actually finding the right index for a given pair will inevitably involve some scan through the list.

Maybe it would be worth retaining these existing `ibd_finder_find_sample_pair_index` functions for whenever the user does just want to calculate IBD for all sample pairs (which, as I said in the call, I think would be a sensible default to assume)? This linear scanning for pair indices is going to be suboptimal in exactly that case.